control_apps/src/main.rs

```diff
@@ -0,0 +1,409 @@
+use std::fmt;
```

Comment:
Initial comments:

- [ ]  Try breaking this up into a couple different files.  What I can see from a glance are three distinct candidates: Valves, Control Manager, and Tests
- [ ] Add your tests/build check to the `.github/workflows/rust.yml` in the base of the directory if you have time.